{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping UIM-reads using Novoalign\n",
    "1. input fastq files and unmapped bam (u.bam) files containing unique molecular index (UIM, aka \"molelular barcode\") information were processed by GATK UIM workflow\n",
    "1. mapping using Novoalign\n",
    "1. sequence read name-wise sorting using Novosort\n",
    "1. obtain marged.bam files from seqsort.bam and u.bam files  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /peta/top/genetics/IRUD-P/AnalysesPrimary/NGSK_IRUD_dummy_dummy/run190909/Sample_NGSK_IRUD_dummy_dummy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "合計 50893720\n",
      "drwxrwsr-x 2 misshie genetics       4096  9月 19 23:54 \u001b[0m\u001b[38;5;5m.\u001b[0m\n",
      "drwxrwsr-x 3 misshie genetics       4096  9月 19 15:44 \u001b[38;5;5m..\u001b[0m\n",
      "lrwxrwxrwx 1 misshie genetics        160  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001.u.bam\u001b[0m -> /peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001.u.bam\n",
      "lrwxrwxrwx 1 misshie genetics        170  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R1_001.fastq.gz\u001b[0m\u001b[K -> \u001b[01;31m/peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R1_001.fastq.gz\u001b[0m\u001b[K\n",
      "lrwxrwxrwx 1 misshie genetics        170  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R2_001.fastq.gz\u001b[0m\u001b[K -> \u001b[01;31m/peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R2_001.fastq.gz\u001b[0m\u001b[K\n",
      "lrwxrwxrwx 1 misshie genetics        160  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002.u.bam\u001b[0m -> /peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002.u.bam\n",
      "lrwxrwxrwx 1 misshie genetics        170  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002_R1_001.fastq.gz\u001b[0m\u001b[K -> \u001b[01;31m/peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002_R1_001.fastq.gz\u001b[0m\u001b[K\n",
      "lrwxrwxrwx 1 misshie genetics        170  9月 19 15:44 \u001b[38;5;37mNGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002_R2_001.fastq.gz\u001b[0m\u001b[K -> \u001b[01;31m/peta/top/raw_data/HiSeq2500/190909_D00385_0273_BHY2VKBCX2_UMI/Unaligned/Project_IRUD-P/Sample_NGSK_IRUD_dummy_dummy/NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002_R2_001.fastq.gz\u001b[0m\u001b[K\n",
      "-rw-r--r-- 1 misshie genetics 9092655291  9月 19 23:48 NGSK_IRUD_dummy_dummy_L001.unsort.bam\n",
      "-rw-r--r-- 1 misshie genetics       3123  9月 19 23:48 NGSK_IRUD_dummy_dummy_L001.unsort.bam.log\n",
      "-rw-r--r-- 1 misshie genetics     337672  9月 19 23:48 NGSK_IRUD_dummy_dummy_L001.unsort.bam.qcal\n",
      "-rw-r--r-- 1 misshie genetics 8452283236  9月 19 23:54 NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam\n",
      "-rw-r--r-- 1 misshie genetics        841  9月 19 23:54 NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam.log\n",
      "-rw-r--r-- 1 misshie genetics 8836270808  9月 20 00:22 NGSK_IRUD_dummy_dummy_L001.unsort.merged.bam\n",
      "-rw-r--r-- 1 misshie genetics      11329  9月 20 00:22 NGSK_IRUD_dummy_dummy_L001.unsort.merged.bam.log\n",
      "-rw-r--r-- 1 misshie genetics 8879035974  9月 19 23:41 NGSK_IRUD_dummy_dummy_L002.unsort.bam\n",
      "-rw-r--r-- 1 misshie genetics       3126  9月 19 23:41 NGSK_IRUD_dummy_dummy_L002.unsort.bam.log\n",
      "-rw-r--r-- 1 misshie genetics     331840  9月 19 23:41 NGSK_IRUD_dummy_dummy_L002.unsort.bam.qcal\n",
      "-rw-r--r-- 1 misshie genetics 8248392700  9月 19 23:48 NGSK_IRUD_dummy_dummy_L002.unsort.bam.seqsort.bam\n",
      "-rw-r--r-- 1 misshie genetics        841  9月 19 23:48 NGSK_IRUD_dummy_dummy_L002.unsort.bam.seqsort.bam.log\n",
      "-rw-r--r-- 1 misshie genetics 8605751376  9月 20 00:18 NGSK_IRUD_dummy_dummy_L002.unsort.merged.bam\n",
      "-rw-r--r-- 1 misshie genetics      11320  9月 20 00:18 NGSK_IRUD_dummy_dummy_L002.unsort.merged.bam.log\n",
      "-rw-r--r-- 1 misshie genetics          0  9月 19 15:48 NovoNGSK_IRUD_dummy_dummy-0.o1078364\n",
      "-rw-r--r-- 1 misshie genetics          0  9月 19 15:48 NovoNGSK_IRUD_dummy_dummy-1.o1078365\n",
      "-rw-rw-r-- 1 misshie genetics       2269  9月 19 15:44 submit-Novoalign000.sh\n",
      "-rw-rw-r-- 1 misshie genetics       2269  9月 19 15:44 submit-Novoalign001.sh\n"
     ]
    }
   ],
   "source": [
    "ls -al"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#$ -S /bin/bash\n",
      "#$ -N NovoNGSK_IRUD_dummy_dummy-0\n",
      "#$ -cwd\n",
      "#$ -m abe\n",
      "#$ -M root\n",
      "#$ -j yes\n",
      "#$ -V\n",
      "#$ -pe smp 32\n",
      "\n",
      "NumProcess=5\n",
      "CorePerProcessor=8\n",
      "NovoMPI=/xcatopt/novocraft-v3.03.02/novoalignMPI\n",
      "Samtools=/xcatopt/samtools-1.2/samtools\n",
      "Novoindex=/scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix\n",
      "suffix=\"__TEMP__\"\n",
      "\n",
      "\n",
      "mpiexec \\\n",
      "    -np ${NumProcess} \\\n",
      "    ${NovoMPI} \\\n",
      "    -c${CorePerProcessor} \\\n",
      "    --mmapoff \\\n",
      "    --hugePage \\\n",
      "    -d ${Novoindex} \\\n",
      "    -f NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R1_001.fastq.gz NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R2_001.fastq.gz \\\n",
      "    -o SAM '@RG\\tID:NGSK_IRUD_dummy_dummy\\tSM:NGSK_IRUD_dummy_dummy\\tLB:NGSK_IRUD_dummy_dummy\\tPL:ILLUMINA' \\\n",
      "    -r Random \\\n",
      "    -k -K NGSK_IRUD_dummy_dummy_L001.unsort.bam.qcal${suffix} \\\n",
      "    --rOQ \\\n",
      "    --rNMOri \\\n",
      "    --matchreward 3 \\\n",
      "    2> NGSK_IRUD_dummy_dummy_L001.unsort.bam.log${suffix} \\\n",
      "    < /dev/null \\\n",
      "    | ${Samtools} view -1 -bS - > NGSK_IRUD_dummy_dummy_L001.unsort.bam${suffix}\n",
      "\n",
      "if [ $? -eq 0 ] ; then\n",
      "    mv NGSK_IRUD_dummy_dummy_L001.unsort.bam${suffix} NGSK_IRUD_dummy_dummy_L001.unsort.bam\n",
      "    mv NGSK_IRUD_dummy_dummy_L001.unsort.bam.qcal${suffix} NGSK_IRUD_dummy_dummy_L001.unsort.bam.qcal\n",
      "    mv NGSK_IRUD_dummy_dummy_L001.unsort.bam.log${suffix} NGSK_IRUD_dummy_dummy_L001.unsort.bam.log\n",
      "fi\n",
      "\n",
      "NumThreads=32\n",
      "scratch=/peta/btm/scratch\n",
      "ram=2G\n",
      "NovoSort=/xcatopt/novocraft-v3.03.02/novosort\n",
      "\n",
      "${NovoSort} \\\n",
      "    --threads ${NumThreads} \\\n",
      "    --tmpdir ${scratch} \\\n",
      "    --ram ${ram} \\\n",
      "    --namesort \\\n",
      "    --keepTags \\\n",
      "    --output NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam \\\n",
      "    NGSK_IRUD_dummy_dummy_L001.unsort.bam \\\n",
      "    > NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam.log 2>&1\n",
      "\n",
      "java=/xcatopt/jre1.8.0_212/bin/java\n",
      "jar=/xcatopt/picard-tools-2.19.0/picard.jar\n",
      "temp=/peta/btm/scratch\n",
      "xmx=24g\n",
      "ref=/scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.fasta\n",
      "\n",
      "${java} \\\n",
      "    -Xmx${xmx} \\\n",
      "    -Djava.io.tmpdir=${temp} \\\n",
      "    -jar ${jar} MergeBamAlignment \\\n",
      "    ALIGNED=NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam \\\n",
      "    UNMAPPED=NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001.u.bam \\\n",
      "    OUTPUT=NGSK_IRUD_dummy_dummy_L001.unsort.merged.bam \\\n",
      "    REFERENCE_SEQUENCE=${ref} \\\n",
      "    VALIDATION_STRINGENCY=LENIENT \\\n",
      "    MAX_INSERTIONS_OR_DELETIONS=-1 \\\n",
      "    SORT_ORDER=unsorted \\\n",
      "    > NGSK_IRUD_dummy_dummy_L001.unsort.merged.bam.log 2>&1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cat submit-Novoalign000.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 misshie genetics 9898265304 12月 26  2013 /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[KchrM\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr1\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr2\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr3\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr4\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr5\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr6\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr7\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr8\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr9\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr10\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr11\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr12\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr13\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr14\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr15\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr16\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr17\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr18\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr19\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr20\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr21\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[Kchr22\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[KchrX\n",
      "\u001b[01;38;5;202m\u001b[K>\u001b[m\u001b[KchrY\n"
     ]
    }
   ],
   "source": [
    "ls -al /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix\n",
    "cat /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.fa | grep '^>'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# novoalignMPI (V3.03.02 - Build Nov 12 2015 @ 14:10:30) - A short read aligner with qualities.\n",
      "# (C) 2008-2015 Novocraft Technologies Sdn Bhd.\n",
      "# License file: /xcatopt/novocraft-v3.03.02/novoalign.lic\n",
      "# Licensed to Department of HUMAN GENETICS, Atomic Bomb Disease Institute, Nagasaki University Graduate School of Biomedical Science\n",
      "#  novoalignMPI -c8 --mmapoff --hugePage -d /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix -f NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R1_001.fastq.gz NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L001_R2_001.fastq.gz -o SAM @RG\\tID:NGSK_IRUD_dummy_dummy\\tSM:NGSK_IRUD_dummy_dummy\\tLB:NGSK_IRUD_dummy_dummy\\tPL:ILLUMINA -r Random -k -K NGSK_IRUD_dummy_dummy_L001.unsort.bam.qcal__TEMP__ --rOQ --rNMOri --matchreward 3 \n",
      "# Starting at Thu Sep 19 15:48:47 2019\n",
      "# Interpreting input files as Sanger FASTQ.\n",
      "# Warning Hugepage allocation for /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix failed, size 9898265304: Cannot allocate memory. Trying malloc()\n",
      "# Warning Hugepage allocation for /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.dbSNP138.ucsc.nix failed, size 9898265304: Cannot allocate memory. Trying malloc()\n",
      "# Index Build Version: 3.2\n",
      "# Hash length: 14\n",
      "# Step size: 2\n",
      "tcmalloc: large alloc 9898270720 bytes == 0x52c6000 @  0x76162a 0x40f2b0 0x41f875 0x413ec9 0x408c2f 0x408d6d 0x404680 0x6a974a 0x4082f1\n",
      "tcmalloc: large alloc 9898270720 bytes == 0x45f0000 @  0x76162a 0x40f2b0 0x41f875 0x413ec9 0x408c2f 0x408d6d 0x404680 0x6a974a 0x4082f1\n",
      "#       Paired Reads: 25473581\n",
      "#       Proper Pairs: 24705144 (97.0%)\n",
      "#     Read Sequences: 50947162\n",
      "#   Unique Alignment: 48899756 (96.0%)\n",
      "#       Multi Mapped:  1978693 ( 3.9%)\n",
      "#   No Mapping Found:    16478 ( 0.0%)\n",
      "# QC Failures...\n",
      "# Homopolymer Filter:       82 ( 0.0%)\n",
      "#        Read Length:    52153 ( 0.1%)\n",
      "#       Elapsed Time: 28806.986 (secs.)\n",
      "#           CPU Time: 14491.44 (min.)\n",
      "# Fragment Length Distribution\n",
      "#\tFrom\tTo\tCount\n",
      "#\t15\t29\t5\n",
      "#\t30\t44\t2966\n",
      "#\t45\t59\t57354\n",
      "#\t60\t74\t305596\n",
      "#\t75\t89\t819655\n",
      "#\t90\t104\t1463770\n",
      "#\t105\t119\t2017036\n",
      "#\t120\t134\t2356131\n",
      "#\t135\t149\t2456434\n",
      "#\t150\t164\t2373113\n",
      "#\t165\t179\t2164000\n",
      "#\t180\t194\t1880164\n",
      "#\t195\t209\t1578749\n",
      "#\t210\t224\t1285548\n",
      "#\t225\t239\t1021512\n",
      "#\t240\t254\t792496\n",
      "#\t255\t269\t607116\n",
      "#\t270\t284\t457985\n",
      "#\t285\t299\t340825\n",
      "#\t300\t314\t251733\n",
      "#\t315\t329\t184760\n",
      "#\t330\t344\t133485\n",
      "#\t345\t359\t98266\n",
      "#\t360\t374\t71905\n",
      "#\t375\t389\t53040\n",
      "#\t390\t404\t38475\n",
      "#\t405\t419\t28747\n",
      "#\t420\t434\t21196\n",
      "#\t435\t449\t16302\n",
      "#\t450\t464\t12383\n",
      "#\t465\t479\t9666\n",
      "#\t480\t494\t7510\n",
      "#\t495\t509\t6193\n",
      "#\t510\t524\t4961\n",
      "#\t525\t539\t4043\n",
      "#\t540\t554\t3261\n",
      "#\t555\t569\t2797\n",
      "#\t570\t584\t2309\n",
      "#\t585\t599\t2039\n",
      "#\t600\t614\t1717\n",
      "#\t615\t629\t1373\n",
      "#\t630\t644\t1246\n",
      "#\t645\t659\t1064\n",
      "#\t660\t674\t917\n",
      "#\t675\t689\t821\n",
      "#\t690\t704\t731\n",
      "#\t705\t719\t619\n",
      "#\t720\t734\t508\n",
      "#\t735\t749\t514\n",
      "#\t750\t764\t445\n",
      "#\t765\t779\t314\n",
      "#\t780\t794\t278\n",
      "#\t795\t809\t250\n",
      "#\t810\t824\t186\n",
      "#\t825\t839\t198\n",
      "#\t840\t854\t157\n",
      "#\t855\t869\t138\n",
      "#\t870\t884\t122\n",
      "#\t885\t899\t107\n",
      "#\t900\t914\t100\n",
      "#\t915\t929\t79\n",
      "#\t930\t944\t67\n",
      "#\t945\t959\t59\n",
      "#\t960\t974\t57\n",
      "#\t975\t989\t47\n",
      "#\t990\t1004\t42\n",
      "#\t1005\t1019\t33\n",
      "#\t1020\t1034\t30\n",
      "#\t1035\t1049\t25\n",
      "#\t1050\t1064\t17\n",
      "#\t1065\t1079\t16\n",
      "#\t1080\t1094\t8\n",
      "#\t1095\t1109\t12\n",
      "#\t1110\t1124\t6\n",
      "#\t1125\t1139\t4\n",
      "# Mean   173, Std Dev  65.3\n",
      "# Done at Thu Sep 19 23:48:54 2019\n"
     ]
    }
   ],
   "source": [
    "cat NGSK_IRUD_dummy_dummy_L001.unsort.bam.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# novosort (V1.03.08) - Build Nov 12 2015 @ 14:13:10\n",
      "# (C) 2012 NovoCraft Technologies Sdn Bhd.\n",
      "# /xcatopt/novocraft-v3.03.02/novosort --threads 32 --tmpdir /peta/btm/scratch --ram 2G --namesort --keepTags --output NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam NGSK_IRUD_dummy_dummy_L001.unsort.bam\n",
      "# License file: /xcatopt/novocraft-v3.03.02/novoalign.lic\n",
      "# Licensed to Department of HUMAN GENETICS, Atomic Bomb Disease Institute, Nagasaki University Graduate School of Biomedical Science\n",
      "# Worker Threads: 32\n",
      "# Process Headers..\n",
      "# File NGSK_IRUD_dummy_dummy_L001.unsort.bam\n",
      "# Sort Phase..\n",
      "#   Sort Buffers: 33\n",
      "# File NGSK_IRUD_dummy_dummy_L001.unsort.bam, 50947162 alignment records\n",
      "# Merge Phase...\n",
      "# 33 memory buffers\n",
      "# 328 temporary segments\n",
      "#  50947162 alignments merged.\n",
      "# Done.\n",
      "#       Elapsed Time: 5.5 (min.)\n",
      "#           CPU Time: 66.2 (min.)\n"
     ]
    }
   ],
   "source": [
    "cat NGSK_IRUD_dummy_dummy_L001.unsort.bam.seqsort.bam.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2019-09-19 14:48:01\tMergeBamAlignment\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MergeBamAlignment -ALIGNED NGSK_IRUD_dummy_dummy_L002.unsort.bam.seqsort.bam -UNMAPPED NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002.u.bam -OUTPUT NGSK_IRUD_dummy_dummy_L002.unsort.merged.bam -REFERENCE_SEQUENCE /scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.fasta -VALIDATION_STRINGENCY LENIENT -MAX_INSERTIONS_OR_DELETIONS -1 -SORT_ORDER unsorted\n",
      "**********\n",
      "\n",
      "\n",
      "14:48:01.459 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/xcatopt/picard-tools-2.19.0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Thu Sep 19 14:48:01 UTC 2019] MergeBamAlignment UNMAPPED_BAM=NGSK_IRUD_dummy_dummy_CAACCACANNNNNNNNNN_L002.u.bam ALIGNED_BAM=[NGSK_IRUD_dummy_dummy_L002.unsort.bam.seqsort.bam] OUTPUT=NGSK_IRUD_dummy_dummy_L002.unsort.merged.bam MAX_INSERTIONS_OR_DELETIONS=-1 SORT_ORDER=unsorted VALIDATION_STRINGENCY=LENIENT REFERENCE_SEQUENCE=/scratch/Genomes/human_hg19_GRCh37/hg19.hg1x.fasta    ADD_PG_TAG_TO_READS=true PAIRED_RUN=true CLIP_ADAPTERS=true IS_BISULFITE_SEQUENCE=false ALIGNED_READS_ONLY=false ATTRIBUTES_TO_REVERSE=[OQ, U2] ATTRIBUTES_TO_REVERSE_COMPLEMENT=[E2, SQ] READ1_TRIM=0 READ2_TRIM=0 ALIGNER_PROPER_PAIR_FLAGS=false PRIMARY_ALIGNMENT_STRATEGY=BestMapq CLIP_OVERLAPPING_READS=true INCLUDE_SECONDARY_ALIGNMENTS=true ADD_MATE_CIGAR=true UNMAP_CONTAMINANT_READS=false MIN_UNCLIPPED_BASES=32 MATCHING_DICTIONARY_TAGS=[M5, LN] UNMAPPED_READ_STRATEGY=DO_NOT_CHANGE VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Thu Sep 19 14:48:01 UTC 2019] Executing as misshie@n01 on Linux 3.13.0-24-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_212-b10; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.19.0-SNAPSHOT\n",
      "INFO\t2019-09-19 14:48:01\tSamAlignmentMerger\tProcessing SAM file(s): [NGSK_IRUD_dummy_dummy_L002.unsort.bam.seqsort.bam]\n",
      "INFO\t2019-09-19 14:48:41\tAbstractAlignmentMerger\tMerged     1,000,000 records.  Elapsed time: 00:00:40s.  Time for last 1,000,000:   38s.  Last read position: chr5:145,631,410\n",
      "INFO\t2019-09-19 14:49:18\tAbstractAlignmentMerger\tMerged     2,000,000 records.  Elapsed time: 00:01:16s.  Time for last 1,000,000:   36s.  Last read position: chr16:19,566,854\n",
      "INFO\t2019-09-19 14:49:52\tAbstractAlignmentMerger\tMerged     3,000,000 records.  Elapsed time: 00:01:51s.  Time for last 1,000,000:   34s.  Last read position: chr6:35,088,699\n",
      "INFO\t2019-09-19 14:50:30\tAbstractAlignmentMerger\tMerged     4,000,000 records.  Elapsed time: 00:02:29s.  Time for last 1,000,000:   37s.  Last read position: chr1:20,650,014\n",
      "INFO\t2019-09-19 14:51:06\tAbstractAlignmentMerger\tMerged     5,000,000 records.  Elapsed time: 00:03:05s.  Time for last 1,000,000:   36s.  Last read position: chr19:9,068,026\n",
      "INFO\t2019-09-19 14:51:43\tAbstractAlignmentMerger\tMerged     6,000,000 records.  Elapsed time: 00:03:42s.  Time for last 1,000,000:   37s.  Last read position: chr1:55,336,953\n",
      "INFO\t2019-09-19 14:52:18\tAbstractAlignmentMerger\tMerged     7,000,000 records.  Elapsed time: 00:04:16s.  Time for last 1,000,000:   34s.  Last read position: chr1:20,656,632\n",
      "INFO\t2019-09-19 14:52:53\tAbstractAlignmentMerger\tMerged     8,000,000 records.  Elapsed time: 00:04:51s.  Time for last 1,000,000:   35s.  Last read position: chr4:74,964,184\n",
      "INFO\t2019-09-19 14:53:28\tAbstractAlignmentMerger\tMerged     9,000,000 records.  Elapsed time: 00:05:27s.  Time for last 1,000,000:   35s.  Last read position: chr12:56,862,436\n",
      "INFO\t2019-09-19 14:54:03\tAbstractAlignmentMerger\tMerged    10,000,000 records.  Elapsed time: 00:06:02s.  Time for last 1,000,000:   35s.  Last read position: chr10:49,367,226\n",
      "INFO\t2019-09-19 14:54:39\tAbstractAlignmentMerger\tMerged    11,000,000 records.  Elapsed time: 00:06:38s.  Time for last 1,000,000:   35s.  Last read position: chr1:9,099,906\n",
      "INFO\t2019-09-19 14:55:16\tAbstractAlignmentMerger\tMerged    12,000,000 records.  Elapsed time: 00:07:14s.  Time for last 1,000,000:   36s.  Last read position: chr3:48,609,345\n",
      "INFO\t2019-09-19 14:55:52\tAbstractAlignmentMerger\tMerged    13,000,000 records.  Elapsed time: 00:07:50s.  Time for last 1,000,000:   35s.  Last read position: chr2:186,668,485\n",
      "INFO\t2019-09-19 14:56:27\tAbstractAlignmentMerger\tMerged    14,000,000 records.  Elapsed time: 00:08:25s.  Time for last 1,000,000:   35s.  Last read position: chr19:805,026\n",
      "INFO\t2019-09-19 14:57:03\tAbstractAlignmentMerger\tMerged    15,000,000 records.  Elapsed time: 00:09:02s.  Time for last 1,000,000:   36s.  Last read position: chr1:3,548,118\n",
      "INFO\t2019-09-19 14:57:41\tAbstractAlignmentMerger\tMerged    16,000,000 records.  Elapsed time: 00:09:39s.  Time for last 1,000,000:   37s.  Last read position: chr9:138,516,082\n",
      "INFO\t2019-09-19 14:58:16\tAbstractAlignmentMerger\tMerged    17,000,000 records.  Elapsed time: 00:10:15s.  Time for last 1,000,000:   35s.  Last read position: chr15:101,115,149\n",
      "INFO\t2019-09-19 14:58:52\tAbstractAlignmentMerger\tMerged    18,000,000 records.  Elapsed time: 00:10:50s.  Time for last 1,000,000:   35s.  Last read position: chr1:32,210,340\n",
      "INFO\t2019-09-19 14:59:29\tAbstractAlignmentMerger\tMerged    19,000,000 records.  Elapsed time: 00:11:28s.  Time for last 1,000,000:   37s.  Last read position: chr16:28,919,846\n",
      "INFO\t2019-09-19 15:00:05\tAbstractAlignmentMerger\tMerged    20,000,000 records.  Elapsed time: 00:12:03s.  Time for last 1,000,000:   35s.  Last read position: chr3:12,976,814\n",
      "INFO\t2019-09-19 15:00:41\tAbstractAlignmentMerger\tMerged    21,000,000 records.  Elapsed time: 00:12:40s.  Time for last 1,000,000:   36s.  Last read position: chr4:111,480,836\n",
      "INFO\t2019-09-19 15:01:18\tAbstractAlignmentMerger\tMerged    22,000,000 records.  Elapsed time: 00:13:16s.  Time for last 1,000,000:   36s.  Last read position: chr14:102,028,556\n",
      "INFO\t2019-09-19 15:01:54\tAbstractAlignmentMerger\tMerged    23,000,000 records.  Elapsed time: 00:13:52s.  Time for last 1,000,000:   36s.  Last read position: chr12:52,099,047\n",
      "INFO\t2019-09-19 15:02:33\tAbstractAlignmentMerger\tMerged    24,000,000 records.  Elapsed time: 00:14:31s.  Time for last 1,000,000:   39s.  Last read position: chr7:47,879,117\n",
      "INFO\t2019-09-19 15:03:14\tAbstractAlignmentMerger\tMerged    25,000,000 records.  Elapsed time: 00:15:12s.  Time for last 1,000,000:   40s.  Last read position: chr16:30,664,657\n",
      "INFO\t2019-09-19 15:03:50\tAbstractAlignmentMerger\tMerged    26,000,000 records.  Elapsed time: 00:15:49s.  Time for last 1,000,000:   36s.  Last read position: chr22:31,679,668\n",
      "INFO\t2019-09-19 15:04:27\tAbstractAlignmentMerger\tMerged    27,000,000 records.  Elapsed time: 00:16:25s.  Time for last 1,000,000:   36s.  Last read position: chr9:110,249,863\n",
      "INFO\t2019-09-19 15:05:02\tAbstractAlignmentMerger\tMerged    28,000,000 records.  Elapsed time: 00:17:00s.  Time for last 1,000,000:   34s.  Last read position: chr17:74,289,903\n",
      "INFO\t2019-09-19 15:05:39\tAbstractAlignmentMerger\tMerged    29,000,000 records.  Elapsed time: 00:17:37s.  Time for last 1,000,000:   37s.  Last read position: chr15:52,155,194\n",
      "INFO\t2019-09-19 15:06:16\tAbstractAlignmentMerger\tMerged    30,000,000 records.  Elapsed time: 00:18:14s.  Time for last 1,000,000:   36s.  Last read position: chr16:31,201,021\n",
      "INFO\t2019-09-19 15:06:50\tAbstractAlignmentMerger\tMerged    31,000,000 records.  Elapsed time: 00:18:49s.  Time for last 1,000,000:   34s.  Last read position: chr12:111,052,201\n",
      "INFO\t2019-09-19 15:07:27\tAbstractAlignmentMerger\tMerged    32,000,000 records.  Elapsed time: 00:19:26s.  Time for last 1,000,000:   36s.  Last read position: chr17:37,118,153\n",
      "INFO\t2019-09-19 15:08:03\tAbstractAlignmentMerger\tMerged    33,000,000 records.  Elapsed time: 00:20:01s.  Time for last 1,000,000:   35s.  Last read position: chr15:68,020,218\n",
      "INFO\t2019-09-19 15:08:41\tAbstractAlignmentMerger\tMerged    34,000,000 records.  Elapsed time: 00:20:39s.  Time for last 1,000,000:   38s.  Last read position: chr3:31,496,776\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2019-09-19 15:09:17\tAbstractAlignmentMerger\tMerged    35,000,000 records.  Elapsed time: 00:21:16s.  Time for last 1,000,000:   36s.  Last read position: chr11:109,861,580\n",
      "INFO\t2019-09-19 15:09:53\tAbstractAlignmentMerger\tMerged    36,000,000 records.  Elapsed time: 00:21:51s.  Time for last 1,000,000:   35s.  Last read position: chr11:64,557,014\n",
      "INFO\t2019-09-19 15:10:30\tAbstractAlignmentMerger\tMerged    37,000,000 records.  Elapsed time: 00:22:28s.  Time for last 1,000,000:   36s.  Last read position: chr16:855,705\n",
      "INFO\t2019-09-19 15:11:06\tAbstractAlignmentMerger\tMerged    38,000,000 records.  Elapsed time: 00:23:05s.  Time for last 1,000,000:   36s.  Last read position: chr12:1,910,809\n",
      "INFO\t2019-09-19 15:11:40\tAbstractAlignmentMerger\tMerged    39,000,000 records.  Elapsed time: 00:23:38s.  Time for last 1,000,000:   33s.  Last read position: chr19:39,396,119\n",
      "INFO\t2019-09-19 15:12:14\tAbstractAlignmentMerger\tMerged    40,000,000 records.  Elapsed time: 00:24:13s.  Time for last 1,000,000:   34s.  Last read position: chr20:60,893,472\n",
      "INFO\t2019-09-19 15:12:49\tAbstractAlignmentMerger\tMerged    41,000,000 records.  Elapsed time: 00:24:48s.  Time for last 1,000,000:   34s.  Last read position: chr1:161,643,650\n",
      "INFO\t2019-09-19 15:13:24\tAbstractAlignmentMerger\tMerged    42,000,000 records.  Elapsed time: 00:25:23s.  Time for last 1,000,000:   35s.  Last read position: chr5:38,903,948\n",
      "INFO\t2019-09-19 15:14:02\tAbstractAlignmentMerger\tMerged    43,000,000 records.  Elapsed time: 00:26:00s.  Time for last 1,000,000:   37s.  Last read position: chr3:45,773,534\n",
      "INFO\t2019-09-19 15:14:37\tAbstractAlignmentMerger\tMerged    44,000,000 records.  Elapsed time: 00:26:35s.  Time for last 1,000,000:   35s.  Last read position: chr9:66,516,930\n",
      "INFO\t2019-09-19 15:15:14\tAbstractAlignmentMerger\tMerged    45,000,000 records.  Elapsed time: 00:27:13s.  Time for last 1,000,000:   37s.  Last read position: chr13:109,779,785\n",
      "INFO\t2019-09-19 15:15:51\tAbstractAlignmentMerger\tMerged    46,000,000 records.  Elapsed time: 00:27:50s.  Time for last 1,000,000:   37s.  Last read position: chr1:152,281,346\n",
      "INFO\t2019-09-19 15:16:28\tAbstractAlignmentMerger\tMerged    47,000,000 records.  Elapsed time: 00:28:26s.  Time for last 1,000,000:   36s.  Last read position: chr17:74,944,686\n",
      "INFO\t2019-09-19 15:17:05\tAbstractAlignmentMerger\tMerged    48,000,000 records.  Elapsed time: 00:29:03s.  Time for last 1,000,000:   36s.  Last read position: chr17:40,698,845\n",
      "INFO\t2019-09-19 15:17:41\tAbstractAlignmentMerger\tMerged    49,000,000 records.  Elapsed time: 00:29:40s.  Time for last 1,000,000:   36s.  Last read position: chr2:240,985,099\n",
      "INFO\t2019-09-19 15:18:18\tAbstractAlignmentMerger\tMerged    50,000,000 records.  Elapsed time: 00:30:16s.  Time for last 1,000,000:   36s.  Last read position: chr6:155,743,890\n",
      "INFO\t2019-09-19 15:18:22\tAbstractAlignmentMerger\tWrote 50033750 alignment records and 2527656 unmapped reads.\n",
      "[Thu Sep 19 15:18:22 UTC 2019] picard.sam.MergeBamAlignment done. Elapsed time: 30.35 minutes.\n",
      "Runtime.totalMemory()=2250244096\n"
     ]
    }
   ],
   "source": [
    "cat NGSK_IRUD_dummy_dummy_L002.unsort.merged.bam.log"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
